pcdsdevices/positioner.py

```diff
@@ -0,0 +1,187 @@
+import time
+
+import numpy as np
+from ophyd.positioner import SoftPositioner
+from ophyd.signal import EpicsSignal, Signal
+from ophyd.utils import InvalidState
+
+from .interface import FltMvInterface
+from .utils import schedule_task
+
+
+class FuncPositioner(FltMvInterface, SoftPositioner):
+    """
+    Class for hacking together a positioner-like object.
+
+    Before you use this, make sure your use case cannot be easily handled by a
+    `PVPositioner` or by a `PseudoPositioner`.
+
+    This should be fast to set up, but has the following limitations:
+    - Inspection tools will not work properly, in general
+    - typhos, happi, and similar tools will not work
+    - It is not possible to know ahead of time if any control layer signals
+      poked by this device are connected.
+    - Metadata about the control layer will not be provided.
+    - It's not possible to intelligently subscribe to control layer signals.
+      Everything will be periodic polls on background threads.
+    - Session performance may be negatively impacted by slow polling functions.
+
+    Parameters
+    ----------
+    name : str, keyword-only
+        The name to use when we refer to this positioner.
+
+    move : func, keyword-only
+        The function to call to cause the device to move.
+        The signature must be def fn(position)
+
+    get_pos : func, keyword-only
+        The function to call to check the device's position.
+        The signature must be def fn()
+
+    set_pos : func, optional, keyword-only
+        If provided, the function to call to change the device's shown
+        position without moving it. The signature must be def fn(position)
+
+    stop : func, optional, keyword-only
+        If provided, the function to call to stop the motor.
+        The signature must be def fn()
+
+    done : func, optional, keyword-only
+        If provided, the function to call to check if the device
+        is done moving. The signature must be def fn() -> bool, and it
+        must return True when motion is complete and false otherwise.
+
+    check_value : func, optional, keyword-only
+        If provided, an extra function to call to check if it is safe to
+        move. The signature must be def fn(position) -> raise ValueError
+
+    info : func, optional, keyword-only
+        If provided, an extra function to add data to the ipython console
+        pretty-print. The signature must be def fn() -> dict{str: value}.
+
+    egu : str, optional
+        If provided, the metadata units for the positioner.
+
+    limits : tuple of floats, optional
+        If provided, we'll force all moves to be within these bounds.
```

Comment:
Sounds like it could mean "clamp moves to `limits`" - but without reading further, I'll assume it means "reject values outside of this range"

Reply:
Yes, I'll revise the docs here